src/typemap.c

```diff
@@ -461,18 +462,16 @@ static int jl_typemap_intersection_array_visitor(struct jl_ordereddict_t *a, jl_
             t = ml.node->key;
         }
         else {
-            t = jl_field_type(ml.leaf->sig, offs);
+            t = jl_field_type(jl_unwrap_unionall((jl_value_t*)ml.leaf->sig), offs);
             if (tparam)
                 t = jl_tparam0(t);
         }
         if (ty == (jl_value_t*)jl_any_type || // easy case: Any always matches
             (tparam ?  // need to compute `ty <: Type{t}`
              (jl_is_uniontype(ty) || // punt on Union{...} right now
               jl_typeof(t) == ty || // deal with kinds (e.g. ty == DataType && t == Type{t})
-              (jl_is_type_type(ty) && (jl_is_typevar(jl_tparam0(ty)) ?
-                                       jl_subtype(t, ((jl_tvar_t*)jl_tparam0(ty))->ub, 0) : // deal with ty == Type{<:T}
-                                       jl_subtype(t, jl_tparam0(ty), 0)))) // deal with ty == Type{T{#<:T}}
-                    : jl_subtype(t, ty, 0))) // `t` is a leaftype, so intersection test becomes subtype
+              jl_isa(t, ty)) // deal with ty == Type{T}
```

Comment:
maybe we should just let type intersection always handle this?
```julia
(tparam ? jl_isa(t, ty) : jl_subtype(t, ty))
```
(the comment above is backwards, it actually needs to compute `Type{t} <: ty`, where `t` is a leaftype.)